### PR TITLE
fix: dragging fails for collapsed blocks with Icons, which have been …

### PR DESCRIPTION
**Pull request by tweini**
…deserialized from JSON

## The basics
- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [x] I ran `npm run format` and `npm run lint`

## The details
Fix for "Dragging fails to complete for collapsed procedure blocks that have been deserialised from JSON #6076".
The bug affect every block with icons, not only procedures. So comments in a collapsed block will trigger this bug too.
Precondition is a deserialization from JSON.

### Resolves
[bug #6076](https://github.com/google/blockly/issues/6076)

### Proposed Changes
Provides an appropriate check if the Icon is visible, as this guarantees an initialized state of the Icon Object belonging to the block.
Not moving the not visible Icons should not have side effects, as expanding the block will render the icons so they get a proper state.
An alternative solution would be a simple null check, but this could chew future errors. The specific checks are imho the better way.

### Test Coverage
Manually tested with some blocks on a playground.

Tested on: This bug is not related to the client for obvious reasons.

### Documentation
There is a code comment in the PR explaining the changes and mentions the # of the issue

### CODE REVIEW

**Comment by BeksOmega** [comment]
Hello! Thank you for the contribution :D Were you able to find out why this was happening with JSON but not with XML?

**Comment by tweini** [comment]
Hi Beka! Well, although I am a big fan of cereals, I didn't read your JSO serializer code, yet. Shall I?

**Comment by tweini** [comment]
Okidoki, I managed to track down the bug in the JSO serializer. It misses to call Icon.computeIconLocation() [fyi: all calls to this method in entire codebase](https://github.com/google/blockly/search?q=computeIconLocation)

Some call stacks:

**Call stack when deserialized from xml:**
computeIconLocation (icon.js:176)
moveConnections (block_svg.js:800)
moveBy (block_svg.js:473)
domToWorkspace (xml.js:471)
load (playground.html:173)

**Call stack when deserialized from JSON:**
initBlock (blocks.js:618)
appendPrivate (blocks.js:400)
appendInternal (blocks.js:345)
append (blocks.js:308)
load (blocks.js:683)
load (workspaces.js:89)
load (playground.html:170)

And now the fun begins. Here I would place the computeIconLocation()-calls for the icons:

https://github.com/google/blockly/blob/d2329f8cba661999cc7f03fcd6c0196770b05cdc/core/serialization/blocks.js#L513

#2105 tells about an Icon API. Is there anything in progress?

So, before providing a new PR, I thought it will be better to sort out what is the best way to do it.

I am fine with contributing, and I am happy digging into the Blockly codebase 😀

**Comment by tweini** [comment]
Pleasure is on my side @BeksOmega . Debugging is so much fun 😊 so I did further investigation.

The truth is not as simple as you proposes 😨 , so sorry.

Deserializing an Icon does not set proper location, so the .iconXY_ property remains _null_ although the the icon object is created and attached to the block.

Setting the icon location is a side effect on both serializers. While xml deserialization, the icon is drawn (really appears on the workspace!) without location here:
https://github.com/google/blockly/blob/d2329f8cba661999cc7f03fcd6c0196770b05cdc/core/block_svg.js#L290

This happens no matter the block is collapsed or not. If not collapsed, the icon gets its position with the call from the drawer L337 and is then properly drawn at its new position.

The uncollapsed block draws the icon as mentioned above. But its visibility state is maintained afterwards in block_svg L679 and then the drawer L111 hides it by setting display:none. The location at this point is _null_ and it is set (as a side effect) in block_svg L800.

I am sure you are aware of the problems, as I visited two issues from you during my investigation. For now, this bug must be patched, and a proper call to .computeIconLocation() would raise awareness for the problem in the JSO serializer.

The handling of Icons causes horrible things (citation from #2105 😄 ), which can't be unseen. And I think it is touching issue #5146 as well as .initSvg() does the mentioned call (see above, L290).

**Comment by BeksOmega** [comment]
So what solution are you thinking of @tweini ?

**Comment by tweini** [comment]
> I am sure you are aware of the problems, as I visited two issues from you during my investigation. For now, this bug must be patched, and a proper call to .computeIconLocation() would raise awareness for the problem in the JSO serializer.

Did change my PR with the new solution mentioned above. Hope my shared infos helps. I got a better understanding of the internals.

**Comment by BeksOmega** [comment]
> > I am sure you are aware of the problems, as I visited two issues from you during my investigation. For now, this bug must be patched, and a proper call to .computeIconLocation() would raise awareness for the problem in the JSO serializer.
> 
> Did change my PR with the new solution mentioned above. Hope my shared infos helps. I got a better understanding of the internals.

Sweet! I couldn't tell if by:

> The truth is not as simple as you proposes 😨 , so sorry.

You meant it also needed a different solution hehe. Either way I really appreciate the added context about the internals!

File: core/block_dragger.js [hunk 1]
```diff
@@ -429,7 +429,11 @@ const BlockDragger = class {
     // Moving icons moves their associated bubbles.
     for (let i = 0; i < this.dragIconData_.length; i++) {
       const data = this.dragIconData_[i];
-      data.icon.setIconLocation(Coordinate.sum(data.location, dxy));
+      // Don't move hidden Icons as they can have not been initialized.
+      // fixes #6076
+      if (!this.draggingBlock_.isCollapsed() || !data.icon.collapseHidden) {
```
Comment (BeksOmega) [inline core/block_dragger.js:434]: Do we still need this fix?

Comment (tweini) [reply core/block_dragger.js:434]: Nope, core/block_dragger.js remains untouched. I thought i got rid of the changes I made in my first commit 🤔

Comment (tweini) [reply core/block_dragger.js:434]: Wait! This code did trick my manual testing into false positive! More info in main thread... .

**Comment by tweini** [comment]
As mentioned, I thought this code was removed. The fix fails if the icon is a mutator, as the mutator isn't serialized in the icon state of the serialization object. so we have to move it to the end. PR follows...

**Comment by BeksOmega** [comment]
Heyo! So I think the issue with the build is that icons don't exist in headless mode (as discussed in https://github.com/google/blockly/issues/2105#issuecomment-885948249) If you move the `computeIconLocation` calls to [after this line](https://github.com/google/blockly/blob/d2329f8cba661999cc7f03fcd6c0196770b05cdc/core/serialization/blocks.js#L625) I think it should work.

Thanks for catching that issue with mutators! I wouldn't have seen that from looking at the code.

**Comment by tweini** [comment]
Thanks for your suggestion, works like a charm. I ran the tests locally, and gladly the erroneous commit put this step back in scope 👓 . And I begin understanding what #2105 is about, as it is also referenced in you project cereal doc.
Hope all is fine, the next bug will be fixed later on. Was another deep dive into serialization internals 🤿

**Review by BeksOmega — APPROVED** [review]

**Comment by BeksOmega** [comment]
Thank you again for the fix and all of the digging! :D

**Comment by tweini** [comment]
@BeksOmega I am very happy 😃 , short celebration 🥳 , opened next pull request to prepare more happiness 😆 (and parties 🎉 )

**Pull request merged**